### ISSUE TITLE
Errors of type lstat: no such file or directory

### ISSUE THREAD

**Issue by zcalusic** [restic/restic#692]
Got it several times, typically from /proc entries:
```
error for /proc/4906: lstat /proc/4906: no such file or directory
```

2 observations:

- This is also in warning class, should not be an error! If file has vanished during the backup, then oh well, there's nothing to backup, it's no error, but can be a warning (just like other cases of fast moving files).
- If the file that has vanished is excluded, this is my case for the example above - I have excluded /proc/*, then I think the most appropriate action is to not even warn, just skip over the error.

That lstat() is done on excluded entries possibly uncovers another bug, maybe exclude filter should be more efficient? Pending further investigation...

**Comment by fd0** [comment]
Thanks for reporting this. I don't understand why this is happening, how did you call restic exactly?

**Comment by zcalusic** [comment]
```restic backup --exclude-file=/usr/local/etc/restic-exclude.txt /```

And exclude file contents are like this:  
```
/mnt/*
/proc/*
/run/*
/sys/*
/tmp/*
...
```

**Comment by fd0** [comment]
Interesting, I'll have a look. For now, you can either use `restic backup --one-file-system` (which will stay on the file systems of the directories to be saved) or remove the `*` in each line.

**Comment by fd0** [comment]
What happens here is that restic first reads the entries of `/proc` (because `/proc` itself is not excluded, just the contents of it). It then calls `lstat()` for each entry, and only then consults the filtering rules. When the entry `/proc/4906` vanishes between the first and the second step, the observed error occurs.

Do you have any suggestion what restic can do to prevent this from happening? My only idea is to instruct users to exclude `/proc` instead of `/proc/*`, but since those have slightly different semantics (the latter will backup the directory `/proc` itself, just without the content) there's nothing restic can do automatically...

**Comment by zcalusic** [comment]
You are right that I use /proc/* nomenclature to keep /proc directory itself. This allows disaster recovery, without needed to recreate missing directories after the restore.

I already suggested what I think is the right solution. If lstat(/somedir/somefile) returns a "no such file" error AND /somedir/somefile is excluded, just don't print the error. From the standpoint of the user, it's the expected behaviour. If I don't want to backup some file, I also don't care whether it exists or not. * means ALL files (not starting with .).

**Comment by fd0** [comment]
Hm, I can see where you're coming from. This however requires changing the way the archiver/exclude logic works, quite a lot.

**Comment by MichaelEischer** [comment]
This has been implemented in restic 0.9.3 by checking excludes before calling lstat, see #1912.

**Comment by kmille** [comment]
> Interesting, I'll have a look. For now, you can either use `restic backup --one-file-system` (which will stay on the file systems of the directories to be saved) or remove the `*` in each line (`/proc` instead of `/proc/*`).

I'm still getting this:

```
rclone: 2025/05/06 22:31:20 NOTICE: Config file "/home/.config/rclone/r
using parent snapshot a6998953
load index files
start backup on [/]
error: lstat /proc/4034652: no such file or directory
error: lstat /proc/4034654: no such file or directory
error: lstat /proc/4034655: no such file or directory
error: lstat /proc/4034656: no such file or directory
```

I'm using `--one-file-system`. That was with restic 0.17.
